src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.IO.Windows.cs

```diff
@@ -0,0 +1,193 @@
+            Debug.Assert(_ioPort != 0);
+
+            if (_ioCompletionPollers == null)
+            {
+                EnsureIOCompletionPollers();
+            }
+
+            if (NativeRuntimeEventSource.Log.IsEnabled())
+            {
+                NativeRuntimeEventSource.Log.ThreadPoolIOEnqueue(nativeOverlapped);
+            }
+
+            if (!Interop.Kernel32.PostQueuedCompletionStatus(_ioPort, 0, UIntPtr.Zero, (IntPtr)nativeOverlapped))
+            {
+                ThrowHelper.ThrowApplicationException(Marshal.GetHRForLastWin32Error());
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void EnsureIOCompletionPollers()
+        {
+            _threadAdjustmentLock.Acquire();
+            try
+            {
+                if (_ioCompletionPollers != null)
+                {
+                    return;
+                }
+
+                int pollerCount = (Environment.ProcessorCount - 1) / ProcessorsPerPoller + 1;
+                IOCompletionPoller[] pollers = new IOCompletionPoller[pollerCount];
+                for (int i = 0; i < pollerCount; ++i)
+                {
+                    pollers[i] = new IOCompletionPoller(_ioPort);
+                }
+
+                _ioCompletionPollers = pollers;
+            }
+            catch (Exception ex)
+            {
+                Environment.FailFast("Failed to initialize IO completion pollers.", ex);
+            }
+            finally
+            {
+                _threadAdjustmentLock.Release();
+            }
+        }
+
+        private sealed unsafe class IOCompletionPoller
+        {
+            private const int NativeEventCapacity =
+#if DEBUG
+                32;
+#else
+                1024;
+#endif
+
+            private static readonly Action<Event> ProcessEventDelegate = ProcessEvent;
+
+            private readonly nint _port;
+            private readonly Interop.Kernel32.OVERLAPPED_ENTRY* _nativeEvents;
+            private readonly ThreadPoolTypedWorkItemQueue<Event> _events;
+            private readonly Thread _thread;
+
+            public IOCompletionPoller(nint port)
+            {
+                Debug.Assert(port != 0);
+                _port = port;
+
+                _nativeEvents =
+                    (Interop.Kernel32.OVERLAPPED_ENTRY*)
+                    NativeMemory.Alloc((nuint)NativeEventCapacity * (nuint)Unsafe.SizeOf<Interop.Kernel32.OVERLAPPED_ENTRY>());
+                _events = new ThreadPoolTypedWorkItemQueue<Event>(ProcessEventDelegate);
+
+                // Thread pool threads must start in the default execution context without transferring the context, so
+                // using UnsafeStart() instead of Start()
+                _thread = new Thread(Poll, SmallStackSizeBytes)
+                {
+                    IsThreadPoolThread = true,
+                    IsBackground = true,
+                    Priority = ThreadPriority.Highest,
```

Comment:
This is interesting, I assume it was the same in native code?

Reply:
The polling mechanism in this change is similar to how it's done on Unixes and different from what was done before. Before on Windows, there were proc-count threads (at normal priority) polling for IO completions one completion at a time. In this change, there are just a few threads polling for batches of IO completions. Depending on how frequently and how many IO completions occur, a number of polling threads would be scheduled. These few IO poller threads have to compete with all of the active thread pool worker threads to get a time slice to queue work. Hill climbing typically causes more worker threads to be active than the proc count, and many services configure a much higher min worker thread count than the proc count for other reasons. The more worker threads that are active and scheduled to process work, the more time it will take for IO poller threads to get a time slice. Making them higher priority removes those artificial dependencies. Unfortunately, none of the ASP.NET benchmarks reach 100% CPU utilization, so they would not be able to show the difference.

RE @GSPP:
> Is ThreadPriority.Highest really the way to go? This can freeze a Windows system stopping even system services. This can cause a user or administrator to lose the ability to stop the application as well.

I'm aware of that, these threads are few in number and due to inherent limits I don't think they can starve a system. If these threads get to a point of starving a system, they would reach a point where there would not be any more IO completions to handle and would not be scheduled. For priority I think `AboveNormal` is sufficient, I don't think `Highest` is necessary. Since there's a config setting that can control this I've also added a check to ensure that there would not be too many high-priority IO poller threads.

Reply:
This is probably not directly actionable but at least you're aware that something unexpected might be going on with scheduling.

Reply:
> I wanted to see if a high-priority thread becoming ready will immediately deschedule lower priority threads, or if there is a delay.

I was curious about that too. I was hoping that `Highest` priority at least would deschedule a thread sooner but apparently not, so `AboveNormal` is similar. One poller thread should be sufficient but it's not in Kestrel Plaintext because of this delay, and that's unfortunate because more poller threads add to CPU time.

> The IO dequeing threads might wait for an unexpectedly long time and not actually take precedence.

At the very least they would get first dibs on time ahead of other scheduled threads, that's still a bonus. Without that they could be waiting for a very long time in some cases.

Reply:
@kouvel it is sounds similar to priority inversion and random priority boosts occurs?

Reply:
> it is sounds similar to priority inversion and random priority boosts occurs (as part of way to fight with inversions, which any OS surely doing)?

I didn't quite follow, can you elaborate?

Reply:
In my testing, I disabled priority boosting for the process. It didn't make a difference.

I found that every 5 seconds, Windows will give a few time slices to starved low-priority threads. I found this behavior described on the internet. This is further evidence that scheduling is not done strictly by priority.